src/converters.ts

```diff
@@ -184,53 +94,54 @@ const boolean = new Converter<boolean, boolean>({
 });
 
 export interface DecimalOptions {
-  maxWholeDigits?: number;
-  decimalPlaces?: number;
-  allowNegative?: boolean;
+  maxWholeDigits: number;
+  decimalPlaces: number;
+  allowNegative: boolean;
 }
 
-function decimal(options?: DecimalOptions) {
-  const maxWholeDigits: number =
-    options == null || !options.maxWholeDigits ? 10 : options.maxWholeDigits;
-  const decimalPlaces: number =
-    options == null || !options.decimalPlaces ? 2 : options.decimalPlaces;
-  const allowNegative: boolean =
-    options == null || options.allowNegative == null
-      ? true
-      : options.allowNegative;
-
-  const regex = new RegExp(
-    `^${allowNegative ? `-?` : ``}(0|[1-9]\\d{0,${maxWholeDigits -
-      1}})(\\.\\d{0,${decimalPlaces}})?$`
-  );
-
+function decimal(
+  decimalOptions?:
+    | Partial<DecimalOptions>
+    | ((context: any) => Partial<DecimalOptions>)
+) {
   return new StringConverter<string>({
     emptyRaw: "",
     emptyImpossible: true,
     defaultControlled: controlled.value,
     neverRequired: false,
     preprocessRaw(
       raw: string,
-      options: StateConverterOptionsWithContext
+      converterOptions: StateConverterOptionsWithContext
     ): string {
       raw = raw.trim();
-      return convertSeparators(raw, options);
+      return convertSeparators(raw, converterOptions);
     },
-    rawValidate(raw) {
+    rawValidate(raw, converterOptions) {
       if (raw === "" || raw === ".") {
         return false;
       }
+      if (
+        converterOptions.thousandSeparator === "." &&
+        converterOptions.decimalSeparator == null
+      ) {
+        throw new Error(
+          "thousandSeparator is . while decimalSeparator isn't set. This shouldn't be possible."
```

Comment:
Is it also an error when thousandSeparator and decimalSeparator are both the same value?